src/chainstate/stacks/db/blocks.rs

```diff
@@ -3190,6 +3215,22 @@ impl StacksChainState {
             return Ok(None);
         }
 
+        // NEW in 2.05
+        // if the parent and child epoch differ, then this block cannot descend from a microblock
+        // stream.  Don't store it if this is the case.
+        if StacksChainState::block_crosses_epoch_boundary(
+            db_handle,
+            burn_chain_tip.block_height,
+            block_commit.block_height,
+        )? {
+            if block.has_microblock_parent() {
+                warn!(
+                    "Invalid block, mined in different epoch than parent but confirms microblocks"
+                );
+                return Ok(None);
+            }
+        }
+
```

Comment:
This should probably be documented somewhere in this repo (I'd appreciate a suggestion on _where_), but the way that the 2.05 transition (and other epoch transitions) will work is roughly as follows:

When elected burn height is in an epoch later than its parent:
* The Clarity block connection is initiated with the _parent's_ epoch. In this case, 2.0. That means that the loaded cost functions are `costs.clar`, the block limit is the 2.0 block limit, and the inputs used for cost methods are 2.0 inputs (static sizes for db ops, etc.)
* The `append_block` method invokes `StacksChainState::process_epoch_transition`
* This invokes `ClarityBlockConnection::initialize_epoch_2_05`
* This instantiates the `costs-2` contract, and sets the Clarity epoch version in the Clarity DB
* The cost tracker will still use cost functions set at the beginning of this block (i.e., before `costs-2` was instantiated). The Clarity Tx will still use StacksEpoch = 2.00. The block limit will still be set using 2.00

When the parent block is the first block in a new epoch (i.e., the Stacks _parent_ was a block that applied `initialize_epoch_2_05`):
* The cost tracker will use cost functions of the new epoch
* The ClarityTx will use StacksEpoch = new epoch (so the cost input changes will be applied)
* The block limit will be set using the new epoch
* This is actually the behavior of _every_ block (i.e., all blocks load that information from their Stacks parent), but there's only a change from the behavior of the parent when the parent applied an epoch transition.

Perhaps to make this more concrete, imagine the following scenario:

Epoch 2.05 begins at `n`
Stacks Block A: elected at burn block `n-1`
Stacks Block B: elected at burn block `n`, parent is Stacks Block A
Stacks Block C: elected at burn block `n+1`, parent is Stacks Block A
Stacks Block D: elected at burn block `n+2`, parent is Stacks Block B
Stacks Block E: elected at burn block `n+3`, parent is Stacks Block C
Stacks Block F: elected at burn block `n+4`, parent is Stacks Block D

* Stacks Block A is in epoch 2.0, and evaluates exactly like any other block in 2.0
* Stacks Block B discovers that it is in epoch 2.05, but its Clarity DB hasn't had the epoch change applied. It invokes `initialize_epoch_2_05`. Otherwise, the cost functions applied during B are the epoch 2.0 cost functions, the block limit is a epoch 2.0 block limit.
* Stacks Block C discovers that it is in epoch 2.05, but its Clarity DB hasn't had the epoch change applied. It invokes `initialize_epoch_2_05`. Otherwise, the cost functions applied during C are the epoch 2.0 cost functions, the block limit is a epoch 2.0 block limit.
* Stacks Block D is in epoch 2.05, and its ClarityDB has had the epoch change applied, so it does not invoke `initialize_epoch_2_05`. The cost functions and block limit applied are based on the parent's epoch, which was 2.05.
* Stacks Block E is in epoch 2.05, and its ClarityDB has had the epoch change applied, so it does not invoke `initialize_epoch_2_05`. The cost functions and block limit applied are based on the parent's epoch, which was 2.05.
* Stacks Block F is in epoch 2.05, and its ClarityDB has had the epoch change applied, so it does not invoke `initialize_epoch_2_05`. The cost functions and block limit applied are based on the parent's epoch, which was 2.05.


Given all of that -- how do you think microblocks should be handled? Should we disallow microblocks between A->B, A->C? What about B->D and C->E?

Reply:
Ah, I'd forgotten this.  I thought B and C would have invoked `initialize_epoch_2_05` *and* would use the 2.05 cost functions and block limit?

Reply:
Yep -- I agree. So I think this PR needs to be updated to check the parent's parent and the parent, rather than checking the parent and the current.

Reply:
Okay, this should be addressed now.  I made it so that the chainstate remembers which block(s) encounter an epoch transition, and then requires that any immediate children *not* confirm microblocks.  Then, it's no longer a function of burn block height -- it's solely a function of whether or not the block processed an epoch transition.